# Задание по практике 3

##Обработка пропущенных значений

Обработайте пропущенные значения:


* Удалите пропущенные значения
* Заполните при помощи медианы, среднего для числовых значений
* Заполните при помощи моды для категориальных значений
* Интерполируйте
Выведите статистку после каждого метода и объясните


In [45]:
import pandas as pd
import numpy as np

df = pd.read_csv('heart.csv')
df

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [69]:
def create_holes(dataframe, column_name, hole_percentage = 50):

    if hole_percentage < 0 or hole_percentage > 100:
        raise ValueError("Процент дырявости должен быть в диапазоне от 0 до 100.")

    dataframe_with_holes = dataframe.copy()
    num_holes = int(len(dataframe) * (hole_percentage / 100))
    hole_indices = np.random.choice(len(dataframe), num_holes, replace=False)
    dataframe_with_holes.loc[hole_indices, column_name] = np.nan

    return dataframe_with_holes

df2 = df
df2 = create_holes(df_nan, 'MaxHR')
df2

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    0              0        140          289          0           0   
1     49    1              1        160          180          0           0   
2     37    0              0        130          283          0           1   
3     48    1              2        138          214          0           0   
4     54    0              1        150          195          0           0   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   
915   57    0              2        130          131          0           0   
916   57    1              0        130          236          0           2   
917   38    0              1        138          175          0           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0      NaN               0      0.0         0             0  
1      NaN               0      1.0         1             1  
2      NaN               0      0.0         0             0  
3      NaN               1      1.5         1             1  
4    122.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  
915    NaN               1      1.2         1             1  
916    NaN               0      0.0         1             1  
917    NaN               0      0.0         0             0  

[918 rows x 12 columns]

In [70]:
#Удаление
df_cleaned = df2.dropna()
df_cleaned

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
4     54    0              1        150          195          0           0   
5     39    0              1        120          339          0           0   
7     54    0              0        110          208          0           0   
8     37    0              2        140          207          0           0   
14    42    1              1        115          211          0           1   
..   ...  ...            ...        ...          ...        ...         ...   
903   56    0              0        130          221          0           2   
905   67    0              1        152          212          0           2   
906   55    1              0        132          342          0           0   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
4    122.0               0      0.0         0             0  
5    170.0               0      0.0         0             0  
7    142.0               0      0.0         0             0  
8    130.0               1      1.5         1             1  
14   137.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
903  163.0               0      0.0         0             0  
905  150.0               0      0.8         1             1  
906  166.0               0      1.2         0             0  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  

[227 rows x 12 columns]

In [71]:
"""
Так как у категориальных переменных нет среднего и медианы,
их надо или заменить с помощью моды, или преобразовать в числа
"""

df3 = df.select_dtypes(['object']).columns
df2[df3] = df2[df3].apply( lambda x: pd.factorize (x)[0])

pd.options.display.float_format = '{:.5}'.format

In [72]:
#Заполнение при помощи медианы
df_median = df2.fillna(df2.median())
df_median

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    0              0        140          289          0           0   
1     49    1              1        160          180          0           0   
2     37    0              0        130          283          0           1   
3     48    1              2        138          214          0           0   
4     54    0              1        150          195          0           0   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   
915   57    0              2        130          131          0           0   
916   57    1              0        130          236          0           2   
917   38    0              1        138          175          0           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    135.0               0      0.0         0             0  
1    135.0               0      1.0         1             1  
2    135.0               0      0.0         0             0  
3    135.0               1      1.5         1             1  
4    122.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  
915  135.0               1      1.2         1             1  
916  135.0               0      0.0         1             1  
917  135.0               0      0.0         0             0  

[918 rows x 12 columns]

In [73]:
#Заполнение при помощи среднего
df_filled_mean = df2.fillna(df2.mean())
df_filled_mean

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    0              0        140          289          0           0   
1     49    1              1        160          180          0           0   
2     37    0              0        130          283          0           1   
3     48    1              2        138          214          0           0   
4     54    0              1        150          195          0           0   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   
915   57    0              2        130          131          0           0   
916   57    1              0        130          236          0           2   
917   38    0              1        138          175          0           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0   134.89               0      0.0         0             0  
1   134.89               0      1.0         1             1  
2   134.89               0      0.0         0             0  
3   134.89               1      1.5         1             1  
4    122.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  
915 134.89               1      1.2         1             1  
916 134.89               0      0.0         1             1  
917 134.89               0      0.0         0             0  

[918 rows x 12 columns]

In [74]:
#Заполнение при помощи моды
df_filled_mode = df2.fillna(df["MaxHR"].mode()[0])
"""
Мода возвращает объект Series, так как
значений моды может быть несколько
"""
df_filled_mode

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    0              0        140          289          0           0   
1     49    1              1        160          180          0           0   
2     37    0              0        130          283          0           1   
3     48    1              2        138          214          0           0   
4     54    0              1        150          195          0           0   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   
915   57    0              2        130          131          0           0   
916   57    1              0        130          236          0           2   
917   38    0              1        138          175          0           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    150.0               0      0.0         0             0  
1    150.0               0      1.0         1             1  
2    150.0               0      0.0         0             0  
3    150.0               1      1.5         1             1  
4    122.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  
915  150.0               1      1.2         1             1  
916  150.0               0      0.0         1             1  
917  150.0               0      0.0         0             0  

[918 rows x 12 columns]

In [75]:
#Заполнение при помощи интерполяции
df_interpolated = df2.interpolate()
df_interpolated

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0     40    0              0        140          289          0           0   
1     49    1              1        160          180          0           0   
2     37    0              0        130          283          0           1   
3     48    1              2        138          214          0           0   
4     54    0              1        150          195          0           0   
..   ...  ...            ...        ...          ...        ...         ...   
913   45    0              3        110          264          0           0   
914   68    0              2        144          193          1           0   
915   57    0              2        130          131          0           0   
916   57    1              0        130          236          0           2   
917   38    0              1        138          175          0           0   

     MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0      NaN               0      0.0         0             0  
1      NaN               0      1.0         1             1  
2      NaN               0      0.0         0             0  
3      NaN               1      1.5         1             1  
4    122.0               0      0.0         0             0  
..     ...             ...      ...       ...           ...  
913  132.0               0      1.2         1             1  
914  141.0               0      3.4         1             1  
915  141.0               1      1.2         1             1  
916  141.0               0      0.0         1             1  
917  141.0               0      0.0         0             0  

[918 rows x 12 columns]

**Объясните**

Обработка пропущенных значений требует осторожности и зависит от специфики данных и целей анализа. При замене пропущенных значений средними или медианными значениями может произойти искажение данных, особенно если количество пропущенных значений велико и они распределены неравномерно. Интерполяция также не всегда является точным методом заполнения пропущенных значений, особенно при большом количестве или значительной вариации в данных. Еще одной проблемой может быть использование моды вместо средних или медианных значений, если данные распределены неоднородно и мода не является представительной для заполнения пропущенных значений. 

# Поиск выбросов и аномалий

**Те, кто выполняет задания со * делают поиск выбросов с помощью написанного ранее класса**

Найти выбросы по числовой переменной (использовать метод основанный на среднем и на межквартильном размахе). Объясните результат

Вывести статистику с выбросами и без. Сравните и объясните результат


In [83]:
#Методом среднего

# Функция для определения выбросов методом среднего
def detect_outliers_mean(dataframe, column_name, threshold=3):
    """
    Функция для определения выбросов методом среднего
    Args:
        dataframe: DataFrame, исходный DataFrame
        column_name: str, имя столбца с числовой переменной
        threshold: float, пороговое значение для определения выбросов (по умолчанию: 3)
    Returns:
        DataFrame без выбросов
    """
    mean = np.mean(dataframe[column_name])
    std = np.std(dataframe[column_name])
    lower_bound = mean - threshold * std
    upper_bound = mean + threshold * std
    df_no_outliers = dataframe[(dataframe[column_name] >= lower_bound) & (dataframe[column_name] <= upper_bound)]
    return df_no_outliers

# Вызов функции для поиска и удаления выбросов
df_no_outliers = detect_outliers_mean(df, 'RestingBP')

# Вывод DataFrame без выбросов
df_no_outliers

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[910 rows x 12 columns]

In [86]:
#При помощи межквартильного размаха

def remove_outliers(df, column):
    # Рассчитываем первый и третий квартили
    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)
    
    # Рассчитываем межквартильный размах
    iqr = q3 - q1
    
    # Определяем границы для определения выбросов
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    
    # Фильтруем датафрейм, оставляя только значения внутри границ
    df_filtered = df[(df[column] >= lower_bound) & (df[column] <= upper_bound)]
    
    return df_filtered

# Удаляем выбросы по переменной
df_filtered = remove_outliers(df, 'RestingBP')

df_filtered

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[890 rows x 12 columns]

In [87]:
#Статистика и объяснение

print(f'1. Оригинальный\n{df.describe()}\n\n2. Метод среднего\n{df_no_outliers.describe()}\n\n3. Межквартильный размах\n{df_filtered.describe()}')

1. Оригинальный
         Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease
count  918.0      918.0        918.0      918.0  918.0    918.0         918.0
mean  53.511      132.4        198.8    0.23312 136.81  0.88736       0.55338
std   9.4326     18.514       109.38    0.42305  25.46   1.0666       0.49741
min     28.0        0.0          0.0        0.0   60.0     -2.6           0.0
25%     47.0      120.0       173.25        0.0  120.0      0.0           0.0
50%     54.0      130.0        223.0        0.0  138.0      0.6           1.0
75%     60.0      140.0        267.0        0.0  156.0      1.5           1.0
max     77.0      200.0        603.0        1.0  202.0      6.2           1.0

2. Метод среднего
         Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease
count  910.0      910.0        910.0      910.0  910.0    910.0         910.0
mean  53.493     132.05       199.12    0.23187 136.81  0.88363       0.55055
std   9.4483     17.178      

**Объяснения**

Итак, оригинальный датафрейм содержит некоторые выбросы, но для анализа сводной статистики, были использованы два метода на основе критерия среднего и межквартильного размаха для получения более точных и надежных результатов. В исходном наборе данных у нас есть 918 наблюдений, тогда как в методе среднего и межквартильном размахе мы исключили некоторые выбросы, получив результирующие наборы данных с 910 и 890 наблюдениями соответственно.

# Дубликаты

Проверьте датафрейм на дубликаты и удалите их, если дублей нет - создайте.

In [90]:
#Проверка на дубли
df_dub = df[df.duplicated()]
df_dub

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [91]:
#Дублей нет:( Надо создать
df = pd.concat((df, df.loc[0:100]))
df_dub = df[df.duplicated()]
df_dub

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
96    43   M           ATA        142          207          0     Normal   
97    39   M           NAP        160          147          1     Normal   
98    56   M           ASY        120           85          0     Normal   
99    41   M           ATA        125          269          0     Normal   
100   65   M           ASY        130          275          0         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
96     138              N      0.0       Up             0  
97     160              N      0.0       Up             0  
98     140              N      0.0       Up             0  
99     144              N      0.0       Up             0  
100    115              Y      1.0     Flat             1  

[101 rows x 12 columns]

In [92]:
#Удаление дублей
df = df.drop_duplicates()
df_no_dub = df[df.duplicated()]
df_no_dub

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

## Преобразование признаков

Масштабируйте числовые признаки через MinMaxScaler, StandardScaler

In [94]:
num_features = df.select_dtypes(include=['float64']).columns
num_features

Index(['Oldpeak'], dtype='object')

In [95]:
df.Oldpeak.unique()

array([ 0. ,  1. ,  1.5,  2. ,  3. ,  4. ,  0.5,  2.5,  5. ,  0.8,  0.7,
        1.4,  2.1,  0.4,  0.2,  1.7,  2.2,  0.1,  1.6,  1.3,  0.3,  1.8,
        2.6, -0.9,  2.8, -2.6, -1.5, -0.1,  0.9,  1.1,  2.4, -1. , -1.1,
       -0.7, -0.8,  3.7,  1.2, -0.5, -2. ,  1.9,  3.5,  0.6,  3.1,  2.3,
        3.4,  3.6,  4.2,  3.2,  5.6,  3.8,  2.9,  6.2,  4.4])

In [97]:
from sklearn.preprocessing import MinMaxScaler
# MinMax масштабирование
scaler_minmax = MinMaxScaler()
df_minmax_scaled = df.copy()
df_minmax_scaled[num_features] = scaler_minmax.fit_transform(df[num_features])
df_minmax_scaled.Oldpeak.unique() # преобразованные уникальные значения в числовом призанке в диапозоне [0, 1]

array([0.29545455, 0.40909091, 0.46590909, 0.52272727, 0.63636364,
       0.75      , 0.35227273, 0.57954545, 0.86363636, 0.38636364,
       0.375     , 0.45454545, 0.53409091, 0.34090909, 0.31818182,
       0.48863636, 0.54545455, 0.30681818, 0.47727273, 0.44318182,
       0.32954545, 0.5       , 0.59090909, 0.19318182, 0.61363636,
       0.        , 0.125     , 0.28409091, 0.39772727, 0.42045455,
       0.56818182, 0.18181818, 0.17045455, 0.21590909, 0.20454545,
       0.71590909, 0.43181818, 0.23863636, 0.06818182, 0.51136364,
       0.69318182, 0.36363636, 0.64772727, 0.55681818, 0.68181818,
       0.70454545, 0.77272727, 0.65909091, 0.93181818, 0.72727273,
       0.625     , 1.        , 0.79545455])

In [98]:
from sklearn.preprocessing import StandardScaler
# Standard масштабирование
scaler_std = StandardScaler()
df_std_scaled = df.copy()
df_std_scaled[num_features] = scaler_std.fit_transform(df[num_features])
df_std_scaled.Oldpeak.unique() # преобразованные уникальные значения в числовом призанке в диапозоне [0, 1]

array([-0.83243239,  0.10566353,  0.57471149,  1.04375945,  1.98185537,
        2.91995129, -0.36338443,  1.51280741,  3.85804721, -0.08195566,
       -0.17576525,  0.4809019 ,  1.13756904, -0.45719402, -0.64481321,
        0.76233067,  1.23137863, -0.7386228 ,  0.66852108,  0.3870923 ,
       -0.55100362,  0.85614026,  1.606617  , -1.67671872,  1.79423618,
       -3.27148179, -2.23957627, -0.92624199,  0.01185394,  0.19947312,
        1.41899782, -1.77052831, -1.86433791, -1.48909954, -1.58290913,
        2.63852251,  0.29328271, -1.30148035, -2.70862423,  0.94994986,
        2.45090333, -0.26957484,  2.07566496,  1.32518822,  2.35709374,
        2.54471292,  3.10757047,  2.16947455,  4.42090476,  2.73233211,
        1.88804578,  4.98376231,  3.29518966])

In [99]:
df_scaled = df.copy() # копируем df, чтобы по ссылке не изменять оригинальный
df_scaled[num_features] = df_std_scaled[num_features]
df_scaled

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N -0.83243       Up             0  
1      156              N  0.10566     Flat             1  
2       98              N -0.83243       Up             0  
3      108              Y  0.57471     Flat             1  
4      122              N -0.83243       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N  0.29328     Flat             1  
914    141              N   2.3571     Flat             1  
915    115              Y  0.29328     Flat             1  
916    174              N -0.83243     Flat             1  
917    173              N -0.83243       Up             0  

[918 rows x 12 columns]

Масштабируйте категориальные признаки через OneHotEncoder

In [100]:
# Объявим категориальные признаки

df.Sex = df.Sex.astype('category')
df.ChestPainType = df.ChestPainType.astype('category')
df.RestingECG = df.RestingECG.astype('category')
df.ExerciseAngina = df.ExerciseAngina.astype('category')
df.ST_Slope = df.ST_Slope.astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Age             918 non-null    int64   
 1   Sex             918 non-null    category
 2   ChestPainType   918 non-null    category
 3   RestingBP       918 non-null    int64   
 4   Cholesterol     918 non-null    int64   
 5   FastingBS       918 non-null    int64   
 6   RestingECG      918 non-null    category
 7   MaxHR           918 non-null    int64   
 8   ExerciseAngina  918 non-null    category
 9   Oldpeak         918 non-null    float64 
 10  ST_Slope        918 non-null    category
 11  HeartDisease    918 non-null    int64   
dtypes: category(5), float64(1), int64(6)
memory usage: 62.6 KB


In [101]:
cat_features = df.select_dtypes(include=['category']).columns

df_cat_features = pd.get_dummies(df, columns=cat_features, drop_first=True)

cat_features

Index(['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope'], dtype='object')

In [102]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder()

encode_features = encoder.fit_transform(df[cat_features])

encode_col = encoder.get_feature_names_out(cat_features)
encode_col

array(['Sex_F', 'Sex_M', 'ChestPainType_ASY', 'ChestPainType_ATA',
       'ChestPainType_NAP', 'ChestPainType_TA', 'RestingECG_LVH',
       'RestingECG_Normal', 'RestingECG_ST', 'ExerciseAngina_N',
       'ExerciseAngina_Y', 'ST_Slope_Down', 'ST_Slope_Flat',
       'ST_Slope_Up'], dtype=object)